### PR TITLE
[RFC] remove Requires, GraphMatrices. Add JuMP, MatrixDepot

### PR DESCRIPTION
**Pull request by CarloLucibello**
The package Requires has some well known problems with precompilation, and its use its being discouraged. I myself get some warning and errors every now and then. 

Therefore I'd like to remove Requires from the required packages. We are relying on Requires for JuMP, MatrixDepot and GraphMatrices. Since GraphMatrices is used in just one function wich could be easily integrated back in GraphMatrices itself (@jpfairbanks) and just mentioned in LG docs, I removed it from LG.  The other two packages add to be added to REQUIRE

### CODE REVIEW

**Comment by sbromberger** [comment]
I'd like to see if we can factor out the JuMP and MatrixDepot stuff into their own modules.

**Comment by CarloLucibello** [comment]
What do you mean? MatrixDepot is already used just in Databases...

**Comment by jpfairbanks** [comment]
## Jump
@sbromberger has previously argued for LG being kept small and simple so that other packages can depend on it. require JuMP would go against that principle.
## MatrixDepot
I would like to see a viable plan for an LG core and network of extensions before we decide to include MatrixDepot as a hard dependency.

## Graph Matrices
Before removing GraphMatrices, which would break a lot of my code, we need a plan for how to interoperate the two packages. I am expecting some more development using both packages over the summer and into the fall semester.

GraphMatrices is independent of graph datastructure and should remain so. That means it does not import any functionality from the graph library. If a data structure author wants to get various useful graph matrix features, then they import GraphMatrices and define the necessary constructor CombinatorialAdjacency. The reason GraphMatrices only appears in one function of LightGraphs is because the interoperability was designed to be minimal.

This is a very Julian design. Package A provides functions that work on abstract types, if package B produces a type that matches the necessary interface, then those functions will work on that type. The `CombinatorialAdjacency` function wraps any type representing a graph into a type that behaves as a sparse matrix, and allows you to produce types which match the necessary interface very easily.

`function CombinatorialAdjacency(g::LightGraphs.Graph) ... end` needs to be somewhere, and it can't go into GraphMatrices, because that means GraphMatrices needs to require LightGraphs and we have not made any progress.

File: REQUIRE [hunk 1]
```diff
@@ -1,8 +1,9 @@
 julia 0.4 0.5-
-Requires
 GZip 0.2.18
 StatsBase 0.7.4
 LightXML 0.2.1
 ParserCombinator 1.7.3
 JLD
 Clustering 0.4
+JuMP
```
Comment (sbromberger) [inline REQUIRE]: I am not in favor of this. It violates interdependence guideline 3 (non-core, non-leaf). The LG methods that require JuMP are not integral to the overall package (though their utility / benefit is not at issue).

File: REQUIRE [hunk 2]
```diff
@@ -1,8 +1,9 @@
...
+MatrixDepot
```
Comment (sbromberger) [inline REQUIRE]: This argues for splitting Datasets from LightGraphs out to JuliaGraphs.

**Comment by CarloLucibello** [comment]
ok, I see some valid arguments against this PR, I'll close it. Let's just keep in mind that Requires is not the ideal solution (I've been experiencing some errors, easily solved by closing and opening the REPL again though) and maybe some day we will find some workaround. At least for the time being there is no better solution though

**Comment by sbromberger** [comment]
Requires is definitely NOT the way we want to go. You're absolutely right, @carlolucibello.

**Pull request closed without merging**